File: client/common/protocol/bet_message.go
package protocol

import (
	"encoding/binary"
	"errors"
)

type Bet struct {
	Agency uint32
	Name       string
	Surname    string
	PersonalId string
	Birthdate  string // yyyy-mm-dd

	BetedNumber uint32
}

func (bet *Bet) ShouldAck() bool {
	return true
}

func (bet Bet) makeHeader() []byte {
	length := EXTRA_BET_BYTES +
		len(bet.Name) +
		len(bet.Surname) +
		len(bet.PersonalId) +
		len(bet.Birthdate)

	header := []byte{BET_OP}

	return buildHeader(header, length)
}

func (bet Bet) addField(stream *[]byte, field string) {
	*stream = append(*stream, byte(len(field)))

	*stream = append(*stream, []byte(field)...)
}



func (bet Bet) addBody(stream *[]byte) {
	bet.addField(stream, bet.Name)
	bet.addField(stream, bet.Surname)
	bet.addField(stream, bet.PersonalId)
	bet.addField(stream, bet.Birthdate)

	serializeUint32(stream, bet.BetedNumber)
	serializeUint32(stream, bet.Agency)
}

func (bet *Bet) Serialize() []byte {
	serialized := bet.makeHeader()
	bet.addBody(&serialized)
	return serialized
}

func (bet Bet) deserializeField(fieldLength int, stream *[]byte) (string, error) {
	field := ""
	if len(*stream) < fieldLength {
		return field, errors.New("stream is shorter than expected")
	}

	field = string((*stream)[:fieldLength])

	*stream = (*stream)[fieldLength:]

	return field, nil
}

func (bet Bet) deserializeFieldLength(stream *[]byte) (int, error) {
	if len(*stream) == 0 {
		return 0, errors.New("stream is shorter than expected")
	}
	fieldSize := []byte{0, (*stream)[0]}
	length := binary.BigEndian.Uint16(fieldSize)
	
	*stream = (*stream)[1:]

	return int(length), nil
}

func (bet Bet) getFieldFromStream(stream *[]byte) (string, error) {
	fieldLength, err := bet.deserializeFieldLength(stream)
	field := ""
	if err == nil {
		field, err = bet.deserializeField(fieldLength, stream)
	}

	return field, err

}


func (bet *Bet) Deserialize(stream []byte) (err error) {
	fields := make([]string, 0)
	var field string
	var betedNumber, agency uint32

	err = checkHeader(stream, BET_OP)

	if err != nil {
		return
	}

	stream = stream[HEADER_SIZE:]

	for i := 0; i < 4 && err == nil; i++ {
		field, err = bet.getFieldFromStream(&stream)

		fields = append(fields, field)
	}

	if err == nil {
		betedNumber, err = deserializeUint32(&stream)
		if err != nil {
			return errors.New("invalid bet structure")
		}
		agency, err = deserializeUint32(&stream)
		bet.Agency = agency
		bet.Name = fields[0]
		bet.Surname = fields[1]
		bet.PersonalId = fields[2]
		bet.Birthdate = fields[3]
		bet.BetedNumber = betedNumber
	}

	return err
}

File: server/common/protocol/bet_message.py
from common.protocol.message import Message
from common.utils import Bet


class BetMessage(Message):
    BET_OP = bytes([0x04])
    EXTRA_BET_BYTES = 8
    
    def __init__(self, first_name: str = "", last_name: str = "", document: str = "", birthdate: str = "1970-12-31", number: str = "0"):
        self.bet = Bet('0', first_name, last_name, document, birthdate, number)

    def __eq__(self, __value: object) -> bool:
        try:
            result = self.bet.agency == __value.bet.agency
            result &= self.bet.birthdate == __value.bet.birthdate
            result &= self.bet.document == __value.bet.document
            result &= self.bet.first_name == __value.bet.first_name
            result &= self.bet.last_name == __value.bet.last_name
            result &= self.bet.number == __value.bet.number
        except:
            result = False
        finally:
            return result
        
    def __make_header(self) -> bytes:
        length = self.EXTRA_BET_BYTES +\
            len(self.bet.first_name) +\
            len(self.bet.last_name) +\
            len(self.bet.document) +\
            len(self.bet.birthdate.isoformat())
        
        return self._build_header(self.BET_OP, length)

    def __add_field(self, field: str) -> bytes:
        stream = bytes([len(field)])
        stream += field.encode()
        return stream

    def __add_body(self, stream: bytes) -> bytes:
        stream += self.__add_field(self.bet.first_name)
        stream += self.__add_field(self.bet.last_name)
        stream += self.__add_field(self.bet.document)
        stream += self.__add_field(self.bet.birthdate.isoformat())
        stream += self._serialize_uint32(self.bet.number)

        return stream

    def serialize(self) -> bytes:
        serialized = self.__make_header()
        
        return self.__add_body(serialized)
    
    def __deserialize_field_length(self, stream: bytes) -> int:
        if len(stream) == 0:
            return -1
        return int.from_bytes(stream[:1], self.ENDIAN, signed=False)

    def __deserialize_field(self, field_length: int, stream: bytes) -> (str, int):
        if len(stream) < field_length:
            
            return "", -1
        
        field = stream[:field_length].decode()
        return field, field_length

    def __get_field_from_stream(self, stream: bytes) -> (str, int):
        field_length = self.__deserialize_field_length(stream)
        if field_length < 0:
            
            return "", field_length
        stream = stream[1:]
        field, err = self.__deserialize_field(field_length, stream)
        return field, err

    def deserialize(self, stream: bytes) -> bool:
        if not self._check_header(stream, self.BET_OP):
            
            return False
        
        stream = stream[self.HEADER_SIZE:]
        
        fields = ['0']
        for _ in range(4):
            field, field_size =\
                self.__get_field_from_stream(stream)
            if field_size < 0:
                
                return False
            fields.append(field)
            stream = stream[field_size+1:]

        bet_number = self._deserialize_uint32(stream)
        if bet_number < 0:
            return False
        fields.append(str(bet_number))
        
        self.bet = Bet(*fields)
        
        return True